lib/class-wp-json-comments-controller.php

```diff
@@ -0,0 +1,470 @@
+<?php
+
+/**
+ * Access comments
+ */
+class WP_JSON_Comments_Controller extends WP_JSON_Controller {
+
+	/**
+	 * Get a list of comments.
+	 *
+	 * @param  WP_JSON_Request $request Full details about the request.
+	 * @return WP_Error|WP_JSON_Response|mixed
+	 */
+	public function get_items( $request ) {
+		$prepared_args = $this->prepare_items_query( $request );
+
+		$query = new WP_Comment_Query;
+		$comments = $query->query( $prepared_args );
+
+		foreach ( $comments as &$comment ) {
+			$post = get_post( $comment->comment_post_ID );
+			if ( ! $this->check_read_post_permission( $post ) || ! $this->check_read_permission( $comment ) ) {
+
+				continue;
+			}
+
+			$comment = $this->prepare_item_for_response( $comment, $request );
+		}
+
+		$response = json_ensure_response( $comments );
+
+		return $response;
+	}
+
+	/**
+	 * Get a comment.
+	 *
+	 * @param  WP_JSON_Request $request Full details about the request.
+	 * @return WP_Error|WP_JSON_Response|mixed
+	 */
+	public function get_item( $request ) {
+		$id = (int) $request['id'];
+
+		$comment = get_comment( $id );
+		if ( empty( $comment ) ) {
+			return new WP_Error( 'json_comment_invalid_id', __( 'Invalid comment ID.' ), array( 'status' => 404 ) );
+		}
+
+		if ( ! $this->check_read_permission( $comment ) ) {
+			return new WP_Error( 'json_user_cannot_read', __( 'Sorry, you cannot read this comment.' ), array( 'status' => 401 ) );
+		}
+
+		$post = get_post( $comment->comment_post_ID );
+		if ( empty( $post ) ) {
+			return new WP_Error( 'json_post_invalid_id', __( 'Invalid post ID.' ), array( 'status' => 404 ) );
+		}
+
+		if ( ! $this->check_read_post_permission( $post ) ) {
+			return new WP_Error( 'json_user_cannot_read', __( 'Sorry, you cannot read this post.' ), array( 'status' => 401 ) );
+		}
+
+		$data = $this->prepare_item_for_response( $comment, $request );
+		$response = json_ensure_response( $data );
+
+		return $response;
+	}
+
+	/**
+	 * Create a comment.
+	 *
+	 * @param  WP_JSON_Request $request Full details about the request.
+	 * @return WP_Error|WP_JSON_Response|mixed
+	 */
+	public function create_item( $request ) {
+		if ( ! empty( $request['id'] ) ) {
+			return new WP_Error( 'json_comment_exists', __( 'Cannot create existing comment.' ), array( 'status' => 400 ) );
+		}
+
+		$post = get_post( (int) $request['post_id'] );
+		if ( empty( $post ) ) {
+			return new WP_Error( 'json_post_invalid_id', __( 'Invalid post ID.' ), array( 'status' => 404 ) );
+		}
+
+		if ( ! $this->check_read_post_permission( $post ) ) {
+			return new WP_Error( 'json_user_cannot_read', __( 'Sorry, you cannot read this post.' ), array( 'status' => 401 ) );
```

Comment:
Hmm in the context of `check_read_post_permission` returning a `WP_Error` seems to indicate there was an issue checking the permissions, not that the permission check is 'not allowed'. Pretty easy either way, convention seems to be that it would return `bool`